lib/handler/status.c

```diff
@@ -166,60 +172,43 @@ static void on_req_close(void *p)
-                SEPARATOR ELEMENT("protocol", "%H") SEPARATOR ELEMENT("referer", "%{Referer}i")
-                    SEPARATOR ELEMENT("user-agent", "%{User-agent}i") SEPARATOR
-        /* time */
-        X_ELEMENT("connect-time") SEPARATOR X_ELEMENT("request-header-time") SEPARATOR X_ELEMENT("request-body-time")
-            SEPARATOR X_ELEMENT("request-total-time") SEPARATOR X_ELEMENT("process-time") SEPARATOR X_ELEMENT("response-time")
-                SEPARATOR
-        /* connection */
-        X_ELEMENT("connection-id") SEPARATOR X_ELEMENT("ssl.protocol-version") SEPARATOR X_ELEMENT("ssl.session-reused")
-            SEPARATOR X_ELEMENT("ssl.cipher") SEPARATOR X_ELEMENT("ssl.cipher-bits") SEPARATOR
-        /* http1 */
-        X_ELEMENT("http1.request-index") SEPARATOR
-        /* http2 */
-        X_ELEMENT("http2.stream-id") SEPARATOR X_ELEMENT("http2.priority.received.exclusive")
-            SEPARATOR X_ELEMENT("http2.priority.received.parent") SEPARATOR X_ELEMENT("http2.priority.received.weight")
-                SEPARATOR X_ELEMENT("http2.priority.actual.parent") SEPARATOR X_ELEMENT("http2.priority.actual.weight") SEPARATOR
-        /* misc */
-        ELEMENT("authority", "%V")
-        /* end */
-        "}";
-#undef ELEMENT
-#undef X_ELEMENT
-#undef SEPARATOR
-
-    { /* compile logconf */
-        char errbuf[256];
-        if ((logconf = h2o_logconf_compile(fmt, H2O_LOGCONF_ESCAPE_JSON, errbuf)) == NULL) {
-            h2o_iovec_t resp = h2o_concat(&req->pool, h2o_iovec_init(H2O_STRLIT("failed to compile log format:")),
-                                          h2o_iovec_init(errbuf, strlen(errbuf)));
-            h2o_send_error(req, 400, "Invalid Request", resp.base, 0);
-            return 0;
-        }
-    }
-
     { /* construct collector and send request to every thread */
         struct st_h2o_status_context_t *status_ctx = h2o_context_get_handler_context(req->conn->ctx, &self->super);
         struct st_h2o_status_collector_t *collector = h2o_mem_alloc_shared(NULL, sizeof(*collector), on_collector_dispose);
         size_t i;
 
+        memset(collector, 0, sizeof(*collector));
+        for (i = 0; i < req->conn->ctx->globalconf->statuses.size; i++) {
+            h2o_status_handler_t *sh;
+
+            h2o_vector_reserve(&req->pool, &collector->status_ctx, collector->status_ctx.size + 1);
+            sh = &req->conn->ctx->globalconf->statuses.entries[i];
+
+            if (status_list.base) {
+                if (!h2o_contains_token(status_list.base, status_list.len, sh->name.base, sh->name.len, ',')) {
+                    collector->status_ctx.entries[collector->status_ctx.size].active = 0;
+                    goto Skip;
+                }
+            }
+            if (sh->init) {
+                h2o_iovec_t err = {NULL, 0};
+                void *p;
+                p = sh->init(&err);
+                if (!p) {
+                    h2o_send_error_400(req, "Invalid Request", err.base, 0);
```

Comment:
Unless there is a specific reason, I believe we should not allow `init` callback to fail.

They should be implemented to always succeed (and that makes the code that calls the handlers simpler), and in case there is a chance of a init callback to fail, then the status handler should retain the error within the handler and emit a error message as part of the JSON when its `final` callback is called.